### ISSUE TITLE
[Inference Optimization] Implement Graph Optimization and Operator Fusion

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#409]
## Problem
MISSING: Graph-level optimizations that fuse operators and optimize computational graphs for faster inference.

## Existing
- Issue #280: ONNX export (complementary)
- Issue #277: Some inference optimizations

## Missing Implementations

**Operator Fusion (CRITICAL):**
- Conv + BatchNorm + ReLU fusion
- Matmul + Bias + Activation fusion
- Elementwise operation fusion
- Multi-headed attention fusion

**Graph Optimization (CRITICAL):**
- Constant folding
- Dead code elimination
- Common subexpression elimination
- Layout optimization (NCHW vs NHWC)

**Memory Optimization (HIGH):**
- In-place operations
- Memory reuse
- Gradient checkpointing integration
- Activation memory planning

**Computation Optimization (HIGH):**
- Algebraic simplification
- Strength reduction
- Loop fusion
- Vectorization hints

## Frameworks to Compete With
- TensorRT (NVIDIA)
- TorchScript optimization
- ONNX Runtime optimizations
- TVM/Apache TVM

## Success Criteria
- 2-5x inference speedup from fusion
- Reduced memory footprint
- Integration with existing models
- Benchmarks vs TensorRT/ONNX Runtime
